FeedStoreChallenge/RealmFeedStore.swift

```diff
@@ -0,0 +1,75 @@
+//
+//  RealmFeedStore.swift
+//  FeedStoreChallenge
+//
+//  Created by Romain Brunie on 09/03/2021.
+//  Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+import Foundation
+import RealmSwift
+
+public class RealmFeedStore: FeedStore {
+	private let queue = DispatchQueue(label: "\(RealmFeedStore.self)Queue", qos: .userInitiated, attributes: .concurrent)
+	private let configuration: Realm.Configuration
+		
+	public init(configuration: Realm.Configuration) {
+		self.configuration = configuration
+	}
+	
+	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
+		queue.async(flags: .barrier) {
+			do {
+				let realm = try Realm(configuration: self.configuration)
+				let confObject = realm.objects(RealmFeedCache.self).first
+				if let conf = confObject {
+					try realm.write {
+						realm.delete(conf)
+					}
+				}
+				completion(nil)
+			} catch {
+				completion(error)
+			}
+		}
+	}
+	
+	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
+		queue.async(flags: .barrier) {
+			do {
+				let realm = try Realm(configuration: self.configuration)
+				try realm.write {
+					let feed = feed.map { RealmFeedImage(id: $0.id,
+														 description: $0.description,
+														 location: $0.location,
+														 url: $0.url)
+					}
+					let realmConf = RealmFeedCache(feed: feed, timestamp: timestamp)
+					realm.add(realmConf as Object, update: .modified)
+				}
+				completion(nil)
+			} catch {
+				completion(error)
+			}
+		}
+	}
+	
+	public func retrieve(completion: @escaping RetrievalCompletion) {
+		queue.async {
+			do {
+				let realm = try Realm(configuration: self.configuration)
+				let realmObject = realm.objects(RealmFeedCache.self).first
+				if let conf = realmObject {
+					let feed = try conf.realmFeedtoLocals()
+					completion(.found(feed: feed, timestamp: conf.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch is RealmFeedImage.InvalidFeedImageData {
+				completion(.failure(RealmFeedImage.InvalidFeedImageData()))
+			} catch {
+				completion(.empty)
+			}
```

Comment:
In the case of *any* error, it should complete with `completion(.failure(error))`:

```suggestion
			} catch {
				completion(.failure(error))
			}
```

If the tests are not passing with this logic, the test setup is wrong. Review the test side.

Reply:
hi @caiozullo , If i remove the catch case RealmFeedImage.InvalidFeedImageData, then I have no idea how to fix it because if the configuration url to create Realm is invalid, then I cannot create the Realm instance and I should return empty like the test wants and it is what we did in the course. I am confused with the way to fix this issue

Reply:
Hi @rombiddle, if it fails to create the Realm instance, it should complete with failure (not .empty) because that's an error.

So using an invalid URL is not simulating a 'retrieval error' with Realm. It's simulating a Realm instantiation error. 

In the course we were not using Realm, so you need to find a way to simulate the 'retrieval error' with Realm. If you can't simulate a Realm error, I recommend deleting this test.

Reply:
Hi @caiozullo , I understand I should not return .empty but I feel like if don't handle the noDeletePermissionURL and invalidStoreURL cases for Realm, I am not testing all error cases. For me, testing means I should test everything that can fails, so my app can never crash. In this case if I delete FailableInsertFeedStoreSpecs and FailableDeleteFeedStoreSpecs tests, I am not testing everything.

Reply:
Hi @rombiddle, that's right - you should test all scenarios you can. But the test setup you had in there was not testing the behavior you want. 

If you want to test all the scenarios with `Realm`, check other PRs here to see how other students did it.